service/history/task/transfer_active_task_executor.go

```diff
@@ -337,21 +338,30 @@ func (t *transferActiveTaskExecutor) processCloseExecution(
 
 	// Communicate the result to parent execution if this is Child Workflow execution
 	if replyToParentWorkflow {
-		recordChildCompletionCtx, cancel := context.WithTimeout(ctx, taskRPCCallTimeout)
-		defer cancel()
-		err = t.historyClient.RecordChildExecutionCompleted(recordChildCompletionCtx, &types.RecordChildExecutionCompletedRequest{
-			DomainUUID: parentDomainID,
-			WorkflowExecution: &types.WorkflowExecution{
-				WorkflowID: parentWorkflowID,
-				RunID:      parentRunID,
-			},
-			InitiatedID: initiatedID,
-			CompletedExecution: &types.WorkflowExecution{
-				WorkflowID: task.WorkflowID,
-				RunID:      task.RunID,
-			},
-			CompletionEvent: completionEvent,
-		})
+		targetDomainEntry, err := t.shard.GetDomainCache().GetDomainByID(parentDomainID)
+		if err != nil {
+			return err
+		}
+		if targetCluster, isCrossCluster := t.isCrossClusterTask(task.DomainID, targetDomainEntry); isCrossCluster {
+			// TODO: consider moving this logic to GenerateWorkflowCloseTasks and uxxse here as a back-up to save latency
+			err = t.generateCrossClusterTask(ctx, wfContext, task, targetCluster)
```

Comment:
I think we probably need to do some refactoring here. 
Currently, the impl of `generateCrossClusterTask` will close the transaction for the mutable state and perform a DB update and is intended to be called as the last step when processing a task. It works for the other three types as for them, we just need to generate one new cross-cluster task, and then the task processing is done. 

For close execution, however, it's possible that another task needs to be generated for applying parent close policy and we can't call `generateCrossClusterTask` twice to close the transaction twice and update DB twice (maybe we don't have such check, but I don't think we should do it). Also we still need the mutable state for applying parent close policy.

I guess the idea here is that we only generate the new task but do not update the workflow. Then at the very end of `processCloseExecution`, we will have a complete list of tasks need to be generated, call the update workflow execution method to close the mutable state transaction, and do a DB update.

Similarly, `taskGenerator.GenerateCrossClusterTaskFromTransferTask` probably also need to be refactored as it's possible that we need to generate two cross-cluster tasks from one transfer task (with probably different target cluster), while the current impl assumes a 1:1 relationship. Sorry I didn't plan for that when writing that method.